datafusion/functions-nested/src/array_has.rs

```diff
@@ -232,98 +236,244 @@ fn array_has_inner_for_array(haystack: &ArrayRef, needle: &ArrayRef) -> Result<A
-    }
 
-    Ok(Arc::new(BooleanArray::from(final_contained)))
+        Ok(Arc::new(boolean_builder.finish()))
+    }};
 }
 
-fn array_has_all_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
-    match args[0].data_type() {
-        DataType::List(_) => {
-            array_has_all_and_any_dispatch::<i32>(&args[0], &args[1], ComparisonType::All)
+// String comparison for array_has_all and array_has_any
+macro_rules! array_has_all_and_any_string_internal {
+    ($haystack:expr, $needle:expr, $comparison_type:expr) => {{
+        let mut boolean_builder = BooleanArray::builder($haystack.len());
+        for (arr, sub_arr) in $haystack.iter().zip($needle.iter()) {
+            match (arr, sub_arr) {
+                (Some(arr), Some(sub_arr)) => {
+                    let haystack_array = string_array_to_vec(&arr);
+                    let needle_array = string_array_to_vec(&sub_arr);
+                    boolean_builder.append_value(array_has_string_kernel(
+                        haystack_array,
+                        needle_array,
+                        $comparison_type,
+                    ));
+                }
+                (_, _) => {
+                    boolean_builder.append_null();
+                }
+            }
         }
-        DataType::LargeList(_) => {
-            array_has_all_and_any_dispatch::<i64>(&args[0], &args[1], ComparisonType::All)
+
+        Ok(Arc::new(boolean_builder.finish()))
+    }};
+}
+
+macro_rules! array_has_all_and_any_dispatch {
+    ($haystack:expr, $needle:expr, $comparison_type:expr) => {
+        if $needle.values().is_empty() {
+            let buffer = match $comparison_type {
+                ComparisonType::All => BooleanBuffer::new_set($haystack.len()),
+                ComparisonType::Any => BooleanBuffer::new_unset($haystack.len()),
+            };
+            Ok(Arc::new(BooleanArray::from(buffer)))
+        } else {
+            match $needle.data_type() {
+                DataType::Utf8 | DataType::LargeUtf8 | DataType::Utf8View => {
+                    array_has_all_and_any_string_internal!(
+                        $haystack,
+                        $needle,
+                        $comparison_type
+                    )
+                }
+                _ => general_array_has_for_all_and_any!(
+                    $haystack,
+                    $needle,
+                    $comparison_type
+                ),
+            }
         }
-        _ => exec_err!(
-            "array_has does not support type '{:?}'.",
-            args[0].data_type()
-        ),
-    }
+    };
 }
 
-fn array_has_any_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
-    match args[0].data_type() {
-        DataType::List(_) => {
-            array_has_all_and_any_dispatch::<i32>(&args[0], &args[1], ComparisonType::Any)
+fn array_has_all_and_any_inner(
+    args: &[ArrayRef],
+    comparison_type: ComparisonType,
+) -> Result<ArrayRef> {
+    match (args[0].data_type(), args[1].data_type()) {
```

Comment:
This will generate quite a few different copies of the code (code bloat) -- and I am not sure how important it is to create a specialized version of each type of comparison

Would it be possible instead to update the coercion logic so the argments are always cast to the same type  (it might already be this way, I didn't double check) before `array_has_all` is called?